cmd/admin-handlers.go

```diff
@@ -422,474 +403,245 @@ func (adminAPI adminAPIHandlers) ClearLocksHandler(w http.ResponseWriter, r *htt
+	if !globalIsXL {
+		writeErrorResponseJSON(w, ErrNotImplemented, r.URL)
 		return
 	}
 
-	// Validate request signature.
-	adminAPIErr := checkRequestAuthType(r, "", "", "")
-	if adminAPIErr != ErrNone {
-		writeErrorResponse(w, adminAPIErr, r.URL)
+	bucket, objPrefix, hs, apiErr := extractHealInitParams(r)
+	if apiErr != ErrNone {
+		writeErrorResponseJSON(w, apiErr, r.URL)
 		return
 	}
 
-	// Validate bucket name and check if it exists.
-	vars := r.URL.Query()
-	bucket := vars.Get(string(mgmtBucket))
-	if err := checkBucketExist(bucket, objLayer); err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
-	}
+	// find number of disk in the setup
+	info := objLayer.StorageInfo()
+	numDisks := info.Backend.OfflineDisks + info.Backend.OnlineDisks
 
-	// if dry-run is present in query-params, then only perform validations and return success.
-	if isDryRun(vars) {
-		writeSuccessResponseHeadersOnly(w)
+	nh := newHealSequence(bucket, objPrefix, numDisks, hs)
+	if err := globalAllHealState.LaunchNewHealSequence(nh); err != nil {
+		writeErrorResponseJSON(w, ErrInternalError, r.URL)
+		errorIf(err, "Failed to start heal sequence.")
 		return
 	}
 
-	// Heal the given bucket.
-	err := objLayer.HealBucket(bucket)
+	// Marshal heal path in reply.
+	jsonBytes, err := json.Marshal(struct {
+		HealPath string `json:"healPathId"`
+	}{nh.path})
 	if err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+		writeErrorResponseJSON(w, ErrInternalError, r.URL)
+		errorIf(err, "Failed to marshal JSON for heal start result.")
 		return
 	}
-
-	// Return 200 on success.
-	writeSuccessResponseHeadersOnly(w)
-}
-
-// isDryRun - returns true if dry-run query param was set and false otherwise.
-// otherwise.
-func isDryRun(qval url.Values) bool {
-	if _, dryRun := qval[string(mgmtDryRun)]; dryRun {
-		return true
-	}
-	return false
-}
-
-// healResult - represents result of a heal operation like
-// heal-object, heal-upload.
-type healResult struct {
-	State healState `json:"state"`
+	writeSuccessResponseJSON(w, jsonBytes)
 }
 
-// healState - different states of heal operation
-type healState int
-
-const (
-	// healNone - none of the disks healed
-	healNone healState = iota
-	// healPartial - some disks were healed, others were offline
-	healPartial
-	// healOK - all disks were healed
-	healOK
-)
-
-// newHealResult - returns healResult given number of disks healed and
-// number of disks offline
-func newHealResult(numHealedDisks, numOfflineDisks int) healResult {
-	var state healState
-	switch {
-	case numHealedDisks == 0:
-		state = healNone
+func extractHealStatusParams(r *http.Request) (healPath string, ackIndex int64,
+	errCode APIErrorCode) {
 
-	case numOfflineDisks > 0:
-		state = healPartial
+	vars := mux.Vars(r)
+	bucket, objPrefix := vars["bucket"], vars["prefix"]
+	healPath = bucket + "/" + objPrefix
 
-	default:
-		state = healOK
+	qParams := r.URL.Query()
+	ackIndexStrs := qParams["ackIndex"]
```

Comment:
Is there a better name for this? why not simply `marker`

Reply:
Yes, good idea, it will be more similar to parameters used in AWS S3 APIs. Updated the PR.